crates/script/src/execute.rs

```diff
@@ -0,0 +1,493 @@
+use super::{
+    runner::ScriptRunner,
+    states::{CompiledState, ExecutedState, LinkedState, PreExecutionState, PreSimulationState},
+    JsonResult, NestedValue, ScriptResult,
+};
+use alloy_dyn_abi::FunctionExt;
+use alloy_json_abi::{Function, InternalType, JsonAbi};
+use alloy_primitives::{Address, Bytes, U64};
+use alloy_rpc_types::request::TransactionRequest;
+use async_recursion::async_recursion;
+use ethers_providers::Middleware;
+use eyre::Result;
+use foundry_cli::utils::{ensure_clean_constructor, needs_setup};
+use foundry_common::{
+    fmt::{format_token, format_token_raw},
+    provider::ethers::{get_http_provider, RpcUrl},
+    shell, ContractsByArtifact,
+};
+use foundry_compilers::artifacts::ContractBytecodeSome;
+use foundry_config::{Config, NamedChain};
+use foundry_debugger::Debugger;
+use foundry_evm::{
+    decode::{decode_console_logs, RevertDecoder},
+    inspectors::cheatcodes::{BroadcastableTransaction, BroadcastableTransactions},
+    traces::{
+        identifier::{EtherscanIdentifier, LocalTraceIdentifier, SignaturesIdentifier},
+        render_trace_arena, CallTraceDecoder, CallTraceDecoderBuilder, TraceKind,
+    },
+};
+use futures::future::join_all;
+use itertools::Itertools;
+use std::collections::{HashMap, HashSet};
+use yansi::Paint;
+
+/// Container for data we need for execution which can only be obtained after linking stage.
+pub struct ExecutionData {
+    /// Function to call.
+    pub func: Function,
+    /// Calldata to pass to the target contract.
+    pub calldata: Bytes,
+    /// Bytecode of the target contract.
+    pub bytecode: Bytes,
+    /// ABI of the target contract.
+    pub abi: JsonAbi,
+}
+
+impl LinkedState {
```

Comment:
this should be below type def

Reply:
yeah, I've placed it like this as we had it split across several files before, should I get rid of `states.rs` at all and keep impls of all states along with their definitions?

Reply:
that'd be great, I prefer that style because then you don't need to jump across multiple files when looking at a type:

```rust
struct A

impl A

impl Trait for A

```

although @DaniPopes prefers `impl Trait` above the `impl A` block, I'm indifferent about the order of these two